### DISCUSSION TITLE
Adding custom constraints?

### DISCUSSION THREAD

**Discussion by vuongdanghuy** [google/or-tools#3161] [Routing (and legacy CP) questions] [answered]
Hi everyone.

I'd like to add custom constraints to a VRP. For example I have a data like this:
```python
data = {}
data['initial_routes'] = [
    [1, 9, 2],
    [3, 10, 4]
]
data['pickups_deliveries'] = [
    [5, 7],
    [6, 8],
]
data['num_vehicles'] = 2
data['depot'] = 0
```
I want that if one vehicle is assign all pickups and deliveries then it must visit all pickups consecutively. For example if vehicle 0 is assigned, then it route can be: `1 -> 9 -> 5 -> 6 -> 7 -> 8 -> 2` or `1 -> 5 -> 6 -> 7 -> 9 -> 8 -> 2`. In other word, I think I have to add some constraints like this:
```
if VehicleVar(5) == VehicleVar(6) then
    CumulVar(5) <= CumulVar(7)
    CumulVar(5) <= CumulVar(8)
    CumulVar(6) <= CumulVar(7)
    CumulVar(6) <= CumulVar(8)
```

Anyone has any suggestions?

Thank you.

**Comment by vuongdanghuy** [comment]
UPDATED!

I make a new BoolVar to check if node 5 and 6 are visited by the same vehicle and add the following constraints:
```python
same_vehicle = solver.IsEqualVar(routing.VehicleVar(5), routing.VehicleVar(6))
solver.Add(
    same_vehicle * (time_dimension.CumulVar(8) - time_dimension.CumulVar(5)) >= 0
)
solver.Add(
    same_vehicle * (time_dimension.CumulVar(7) - time_dimension.CumulVar(6)) >= 0
)
```
It makes node 7 and 8 will only be visited after node 5, 6 has been visited. But I still have the case, for example: `5->9->6`.
I've tried the same idea as with node 7 and 8:
```python
solver.Add(
    same_vehicle * (time_dimension.CumulVar(9) - time_dimension.CumulVar(5)) * (time_dimension.CumulVar(9) - time_dimension.CumulVar(6)) >= 0
)
```
but it still see route `5->9->6` as a valid route.
Why? And how I can deal with this case?
Thank you.

**Reply by Mizux** [reply]
so 9 is **before** `min(CumulVar(5), CumuVar(6))` OR **after** `max(CumulVar(5), CumulVar(6))` ?

**Reply by vuongdanghuy** [reply]
That's right, 9 can be before or after but not between. It' different from 7 and 8 because they are delivery nodes so they can only be after. But the constraint for 9: `(CumulVar(5) - CumulVar(9)) * (CumulVar(6) - CumulVar(9)) > 0` does not work.

**Comment by jmarca** [comment]
I am not sure exactly what you are asking and what to have happen, but maybe something like:

```python
for pair in pairs:
    first = manager.NodeToIndex(pair[0])
    second = manager.NodeToIndex(pair[1])
    both_active = routing.ActiveVar(first) * routing.ActiveVar(second)
    same_veh = routing.VehicleVar(first) == routing.VehicleVar(second)

    routing.solver().Add(
        both_active * same_veh * time_dim.CumulVar(first) <
        both_active * same_veh * time_dim.CumulVar(second))
```

There is also a good discussion of FIFO and LIFO policies for the pickup and delivery constraints in the examples [fifo here](https://github.com/google/or-tools/blob/stable/ortools/constraint_solver/samples/vrp_pickup_delivery_fifo.py), [lifo here](https://github.com/google/or-tools/blob/stable/ortools/constraint_solver/samples/vrp_pickup_delivery_lifo.py).  Maybe that is what you want.

**Reply by vuongdanghuy** [reply]
Thanks, I'll take a look at the discussion you said.

What I want is that all pick up node must be visited consecutively. In my example, node 5 and 6 are pickup node and if they are assigned to the same vehicle, let say vehicle 0, then route for vehicle 0 can be:
`1->5->6->9->7->8->2`
or
`1->9->6->5->7->8->2`

But it can't be:
`1->5->9->6->7->8->2` (not valid because 9 is between 5 and 6)
or
`1->5->7->6->9->8->2` (not valid because 7 is between 5 and 6)

I've added 2 constraints to make sure 7 or 8 can't be between 5 and 6, but there's still the case where node 9 is between 5 and 6.

**Comment by jmarca** [comment]
Ah, not what I thought you wanted.

Look at usage of NextVar() in the examples. You can constrain the next value of a node. Maybe search the forum too, as I think I posted there before about that.

**Comment by vuongdanghuy** [comment, marked as answer]
UPDATED!

I found a solution using NextVar and IsMemberVar if anyone needs it. In general case, it can be used if there must be a consecutive route between some nodes (`data['pickups']` in this example).
```python
data['pickups'] = [5, 6]
pickup_indexes = [manager.NodeToIndex(node) for node in data['pickups']]

solver = routing.solver()
for i in range(0, len(data['pickups']) - 1):
    for j in range(i+1, len(data['pickups'])):
        pickup_index0 = manager.NodeToIndex(data['pickups'][i])
        pickup_index1 = manager.NodeToIndex(data['pickups'][j])
        same_vehicle = routing.VehicleVar(pickup_index0) == routing.VehicleVar(pickup_index1)
        cond0 = solver.IsMemberVar(routing.NextVar(pickup_index0), pickup_indexes)
        cond1 = solver.IsMemberVar(routing.NextVar(pickup_index1), pickup_indexes)
        solver.Add(
            same_vehicle * (cond0 + cond1 - 1) >= 0
        )
```
